File: labs/11-lsp-minimum/python/main.py
"""Minimal LSP server over stdio (JSON-RPC with Content-Length framing).

Implements:
  - initialize
  - initialized (notification)
  - textDocument/didOpen
  - textDocument/definition
  - shutdown
  - exit
"""

import argparse
import json
import logging
import re
import sys
from typing import Any

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
    stream=sys.stderr,
)
logger = logging.getLogger("lsp-server")


# ---------- document store ----------

DOCUMENTS: dict[str, str] = {}


def update_document(uri: str, text: str) -> None:
    DOCUMENTS[uri] = text


def find_definition(uri: str, position: dict[str, int]) -> dict[str, Any] | None:
    """Return a Location if the word under the cursor is a function name defined somewhere."""
    text = DOCUMENTS.get(uri, "")
    lines = text.splitlines()
    line_idx = position.get("line", 0)
    char_idx = position.get("character", 0)
    if not (0 <= line_idx < len(lines)):
        return None
    line = lines[line_idx]
    match = re.search(r"[A-Za-z0-9_]+", line[char_idx:] if char_idx < len(line) else "")
    if not match:
        return None
    word = match.group(0)

    # Search all open documents for "def word("
    pattern = re.compile(rf"^def\s+{re.escape(word)}\s*\(", re.MULTILINE)
    for doc_uri, doc_text in DOCUMENTS.items():
        for m in pattern.finditer(doc_text):
            start_line = doc_text[: m.start()].count("\n")
            return {
                "uri": doc_uri,
                "range": {
                    "start": {"line": start_line, "character": 0},
                    "end": {"line": start_line, "character": len(f"def {word}(")},
                },
            }
    return None


# ---------- JSON-RPC framing ----------


def read_message(stream: Any) -> dict[str, Any] | None:
    headers: dict[str, str] = {}
    while True:
        line = stream.readline()
        if not line:
            return None
        line = line.decode("utf-8").strip()
        if not line:
            break
        if ":" in line:
            key, value = line.split(":", 1)
            headers[key.strip().lower()] = value.strip()

    length = int(headers.get("content-length", 0))
    if length == 0:
        return None
    body = stream.read(length)
    return json.loads(body.decode("utf-8"))


def write_message(stream: Any, message: dict[str, Any]) -> None:
    body = json.dumps(message, ensure_ascii=False).encode("utf-8")
    stream.write(f"Content-Length: {len(body)}\r\n\r\n".encode("utf-8"))
    stream.write(body)
    stream.flush()


# ---------- request handlers ----------


def make_response(request_id: Any, result: Any) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": request_id, "result": result}


def make_error(request_id: Any, code: int, message: str) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": request_id, "error": {"code": code, "message": message}}


def handle_request(request: dict[str, Any]) -> dict[str, Any] | None:
    method = request.get("method")
    params = request.get("params", {})
    req_id = request.get("id")
    logger.info("Received %s", method)

    if method == "initialize":
        return make_response(
            req_id,
            {
                "capabilities": {
                    "textDocumentSync": {"openClose": True, "change": 0},
                    "definitionProvider": True,
                },
                "serverInfo": {"name": "agent-labs-lsp", "version": "0.1.0"},
            },
        )

    if method == "initialized":
        return None

    if method == "textDocument/didOpen":
        doc = params.get("textDocument", {})
        update_document(doc.get("uri", ""), doc.get("text", ""))
        return None

    if method == "textDocument/definition":
        td = params.get("textDocument", {})
        pos = params.get("position", {})
        location = find_definition(td.get("uri", ""), pos)
        return make_response(req_id, location)

    if method == "shutdown":
        return make_response(req_id, None)

    if method == "exit":
        return None

    return make_error(req_id, -32601, f"Method not found: {method}")


def serve(stdin=None, stdout=None) -> None:
    stdin = stdin or sys.stdin.buffer
    stdout = stdout or sys.stdout.buffer
    logger.info("LSP server ready on stdio")
    while True:
        try:
            request = read_message(stdin)
        except json.JSONDecodeError as exc:
            logger.error("Bad JSON: %s", exc)
            write_message(stdout, make_error(None, -32700, "Parse error"))
            continue
        if request is None:
            logger.info("EOF reached; shutting down")
            break

        method = request.get("method")
        response = handle_request(request)
        if response is not None:
            write_message(stdout, response)
        if method == "exit":
            break


# ---------- smoke test ----------


def smoke_test():
    import io

    sample_code = '\n'.join([
        'def greet(name):',
        '    return f"Hello, {name}!"',
        '',
        'print(greet("world"))',
    ])
    uri = "file:///tmp/sample.py"

    requests = [
        {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "initialize",
            "params": {"processId": None, "rootUri": None, "capabilities": {}},
        },
        {"jsonrpc": "2.0", "method": "initialized", "params": {}},
        {
            "jsonrpc": "2.0",
            "method": "textDocument/didOpen",
            "params": {"textDocument": {"uri": uri, "languageId": "python", "text": sample_code}},
        },
        {
            "jsonrpc": "2.0",
            "id": 2,
            "method": "textDocument/definition",
            "params": {
                "textDocument": {"uri": uri},
                "position": {"line": 3, "character": 6},  # cursor on greet
            },
        },
        {"jsonrpc": "2.0", "id": 3, "method": "shutdown"},
        {"jsonrpc": "2.0", "method": "exit"},
    ]

    stdin = io.BytesIO()
    for req in requests:
        body = json.dumps(req).encode("utf-8")
        stdin.write(f"Content-Length: {len(body)}\r\n\r\n".encode("utf-8"))
        stdin.write(body)
    stdin.seek(0)

    stdout = io.BytesIO()
    serve(stdin, stdout)
    stdout.seek(0)
    print("Smoke test output:")
    print(stdout.read().decode("utf-8"))


def main():
    parser = argparse.ArgumentParser(description="Minimal LSP server over stdio")
    parser.add_argument(
        "--smoke",
        action="store_true",
        help="Run a self-contained smoke test instead of serving.",
    )
    args = parser.parse_args()
    if args.smoke:
        smoke_test()
    else:
        serve()


if __name__ == "__main__":
    main()

File: labs/11-lsp-minimum/go/main.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"regexp"
	"strconv"
	"strings"
)

var logger = log.New(os.Stderr, "lsp-server: ", log.LstdFlags)

var documents = map[string]string{}

func updateDocument(uri, text string) {
	documents[uri] = text
}

func findDefinition(uri string, position map[string]interface{}) map[string]interface{} {
	text := documents[uri]
	lines := strings.Split(text, "\n")
	lineIdx := 0
	if v, ok := position["line"].(float64); ok {
		lineIdx = int(v)
	}
	charIdx := 0
	if v, ok := position["character"].(float64); ok {
		charIdx = int(v)
	}
	if lineIdx < 0 || lineIdx >= len(lines) {
		return nil
	}
	line := lines[lineIdx]
	if charIdx >= len(line) {
		return nil
	}
	re := regexp.MustCompile(`[A-Za-z0-9_]+`)
	match := re.FindString(line[charIdx:])
	if match == "" {
		return nil
	}
	pattern := regexp.MustCompile(fmt.Sprintf(`^func\s+%s\s*\(`, regexp.QuoteMeta(match)))
	for docURI, docText := range documents {
		for _, m := range pattern.FindAllStringIndex(docText, -1) {
			startLine := strings.Count(docText[:m[0]], "\n")
			return map[string]interface{}{
				"uri": docURI,
				"range": map[string]interface{}{
					"start": map[string]interface{}{"line": startLine, "character": 0},
					"end":   map[string]interface{}{"line": startLine, "character": len(fmt.Sprintf("func %s(", match))},
				},
			}
		}
	}
	return nil
}

// ---------- JSON-RPC framing ----------

func readMessage(r *bufio.Reader) (map[string]interface{}, error) {
	headers := map[string]string{}
	for {
		line, err := r.ReadString('\n')
		if err != nil {
			if err == io.EOF && len(headers) == 0 {
				return nil, io.EOF
			}
			return nil, err
		}
		line = strings.TrimSpace(line)
		if line == "" {
			break
		}
		if idx := strings.Index(line, ":"); idx >= 0 {
			key := strings.ToLower(strings.TrimSpace(line[:idx]))
			value := strings.TrimSpace(line[idx+1:])
			headers[key] = value
		}
	}
	length, _ := strconv.Atoi(headers["content-length"])
	if length == 0 {
		return nil, nil
	}
	body := make([]byte, length)
	_, err := io.ReadFull(r, body)
	if err != nil {
		return nil, err
	}
	var msg map[string]interface{}
	if err := json.Unmarshal(body, &msg); err != nil {
		return nil, err
	}
	return msg, nil
}

func writeMessage(w io.Writer, msg map[string]interface{}) error {
	body, err := json.Marshal(msg)
	if err != nil {
		return err
	}
	_, err = fmt.Fprintf(w, "Content-Length: %d\r\n\r\n%s", len(body), body)
	return err
}

// ---------- request handlers ----------

func makeResponse(requestID interface{}, result interface{}) map[string]interface{} {
	return map[string]interface{}{"jsonrpc": "2.0", "id": requestID, "result": result}
}

func makeError(requestID interface{}, code int, message string) map[string]interface{} {
	return map[string]interface{}{
		"jsonrpc": "2.0",
		"id":      requestID,
		"error":   map[string]interface{}{"code": code, "message": message},
	}
}

func handleRequest(request map[string]interface{}) map[string]interface{} {
	method, _ := request["method"].(string)
	params, _ := request["params"].(map[string]interface{})
	reqID := request["id"]

	logger.Printf("Received %s", method)

	switch method {
	case "initialize":
		return makeResponse(reqID, map[string]interface{}{
			"capabilities": map[string]interface{}{
				"textDocumentSync":   map[string]interface{}{"openClose": true, "change": 0},
				"definitionProvider": true,
			},
			"serverInfo": map[string]interface{}{"name": "agent-labs-lsp", "version": "0.1.0"},
		})
	case "initialized":
		return nil
	case "textDocument/didOpen":
		td, _ := params["textDocument"].(map[string]interface{})
		updateDocument(td["uri"].(string), td["text"].(string))
		return nil
	case "textDocument/definition":
		td, _ := params["textDocument"].(map[string]interface{})
		pos, _ := params["position"].(map[string]interface{})
		return makeResponse(reqID, findDefinition(td["uri"].(string), pos))
	case "shutdown":
		return makeResponse(reqID, nil)
	case "exit":
		return nil
	default:
		return makeError(reqID, -32601, fmt.Sprintf("Method not found: %s", method))
	}
}

func serve(stdin io.Reader, stdout io.Writer) {
	r := bufio.NewReader(stdin)
	w := bufio.NewWriter(stdout)
	logger.Println("LSP server ready on stdio")
	for {
		request, err := readMessage(r)
		if err != nil {
			if err == io.EOF {
				logger.Println("EOF reached; shutting down")
				break
			}
			logger.Printf("Bad JSON: %v", err)
			_ = writeMessage(w, makeError(nil, -32700, "Parse error"))
			_ = w.Flush()
			continue
		}
		if request == nil {
			logger.Println("EOF reached; shutting down")
			break
		}
		method, _ := request["method"].(string)
		response := handleRequest(request)
		if response != nil {
			_ = writeMessage(w, response)
			_ = w.Flush()
		}
		if method == "exit" {
			break
		}
	}
}

// ---------- smoke test ----------

func smokeTest() {
	sampleCode := strings.Join([]string{
		`def greet(name):`,
		`    return f"Hello, {name}!"`,
		"",
		`print(greet("world"))`,
	}, "\n")
	uri := "file:///tmp/sample.py"

	requests := []map[string]interface{}{
		{
			"jsonrpc": "2.0",
			"id":      1,
			"method":  "initialize",
			"params":  map[string]interface{}{"processId": nil, "rootUri": nil, "capabilities": map[string]interface{}{}},
		},
		{"jsonrpc": "2.0", "method": "initialized", "params": map[string]interface{}{}},
		{
			"jsonrpc": "2.0",
			"method":  "textDocument/didOpen",
			"params": map[string]interface{}{
				"textDocument": map[string]interface{}{
					"uri":        uri,
					"languageId": "python",
					"text":       sampleCode,
				},
			},
		},
		{
			"jsonrpc": "2.0",
			"id":      2,
			"method":  "textDocument/definition",
			"params": map[string]interface{}{
				"textDocument": map[string]interface{}{"uri": uri},
				"position":     map[string]interface{}{"line": 3, "character": 6},
			},
		},
		{"jsonrpc": "2.0", "id": 3, "method": "shutdown"},
		{"jsonrpc": "2.0", "method": "exit"},
	}

	var stdin bytes.Buffer
	for _, req := range requests {
		body, _ := json.Marshal(req)
		stdin.WriteString(fmt.Sprintf("Content-Length: %d\r\n\r\n%s", len(body), body))
	}

	var stdout bytes.Buffer
	serve(&stdin, &stdout)
	fmt.Println("Smoke test output:")
	fmt.Println(stdout.String())
}

func main() {
	smoke := flag.Bool("smoke", false, "Run a self-contained smoke test instead of serving.")
	flag.Parse()
	if *smoke {
		smokeTest()
	} else {
		serve(os.Stdin, os.Stdout)
	}
}
